### ISSUE TITLE
Pluggable Auth: Allow OIDC Authentication

### ISSUE THREAD

**Issue by etiennedi** [creativesoftwarefdn/weaviate#801]
as outlined in #628 

## Todos
* [x] regenerate with desired principles (straight on master to minimize impact on others)
* [x] add OIDC option in swagger
* [x] add Authentication switcher to config
  * [x] when no auth is selected, error
  * [x] when only OIDC is enabled, make sure anonymous fails
    * this might require a workaround: https://github.com/go-swagger/go-swagger/issues/1446#issuecomment-475272579
  * [x] when only anonymous is selected, make sure OIDC fails
  * ~~when both anonymous and OIDC is selected, error~~
    * this doesnt make sense, as it is completely viable (depending on AuthZ strategy) to allow both in parallel (e.g. anonymous reads, but only authenticated writes)

* when OIDC is selected
  * [x] validate that all required fields are set in the configuration
  * [x] validate token
  * [x] extract desired claims
  * [x] dev setup (can we programatically spin up a keycloak)
    * we can dump a manually configured keycloak into a json file and import that on startup. That should be sufficient
    * users cannot be exported though, we will have to insert them using the rest api
  * ~~test setup (possibly a second parallel build on CI?)~~
    * since we don't have any valid test-cases that aren't already covered by the unit tests, let's move that to #802. There it will get interesting to test the different rights of different users e2e
  * [x] add option to skip audience validation
* [x] update docs
  * [x] explain selection of auth method
  * [x] outline how each method works
  * [x] reference relevant docs in error messages
